File: Solitaire/Solitaire/wwwroot/js/models/deck.js
class deck {
    constructor() {
        this.cards = new customqueue();
        
    }

    initializedeck() {
        const suits = ['hearts', 'diamonds', 'spade', 'clubs'];
        const ranks = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];
        for (let suit of suits) {
            for (let rank of ranks) {
                this.cards.enqueue(new this.cards(suit, card, false));
            }
        }
    }

    shuffle() {
        let cardsarray = [];
        while (!this.card.isEmpty()) {
            cardsarray.push(this.cards.dequeue());
        }

        for (let i = cardsarray.length - 1; i > 0; i--) {
            const j = math.floor(Math.random() * (i + 1));
            cardsarray[i], cardsarray[j] = cardsarray[j], cardsarray[i];
        }

        for (let card of cardsarray) {
            this.cards.enqueue(card);
        }
    }

    drawcard() {
        return this.cards.isEmpty() ? null : this.cards.dequeue();  
    }

    getsize() {
        return this.cards.size();
    }

}

class GameEngine {
    constructor() {
        this.deck = new deck();

    }
}

File: Solitaire/Solitaire/Models/models/Deck.cs
namespace Solitaire.Models.models;
using System;
using System.Collections.Generic;
using Solitaire.Models.datastructures;

public class Deck
{
    private CustomQueue<Card> cards;

    public Deck()
    {
        cards = new CustomQueue<Card>();
        InitializeDeck();
    }

    public void InitializeDeck()
    {
        string[] suits = { "hearts", "diamonds", "spades", "clubs" };
        int[] ranks = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13 };

        foreach (string suit in suits)
        {
            foreach (int rank in ranks)
            {
                cards.Enqueue(new Card(suit, rank, false));
            }
        }
    }

    public void Shuffle()
    {
        List<Card> cardsArray = new List<Card>();

        // Empty the queue into a list
        while (!cards.IsEmpty())
        {
            cardsArray.Add(cards.Dequeue());
        }

        // Fisher-Yates shuffle
        Random rng = new Random();
        for (int i = cardsArray.Count - 1; i > 0; i--)
        {
            int j = rng.Next(i + 1);
            // Swap cards
            Card temp = cardsArray[i];
            cardsArray[i] = cardsArray[j];
            cardsArray[j] = temp;
        }

        // Put cards back into queue
        foreach (Card card in cardsArray)
        {
            cards.Enqueue(card);
        }
    }

    public Card DrawCard()
    {
        return cards.IsEmpty() ? null : cards.Dequeue();
    }

    public int GetSize()
    {
        return cards.Size();
    }
}